File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
private:
    bool palindrome(int start,int end,string &s){
     while(start<=end){
         if (s[start++] != s[end--])
             return false;         
     }
     return true;
    }
    
    void createPartition(vector<string>&path, vector<vector<string>> &ans,
                         int index, string &s){
        
        if(index >= s.length()){
            ans.emplace_back(path);
            return;
        }        
        for(int i=index;i<s.length();i++){
            if(palindrome(index,i,s)){
                path.emplace_back(s.substr(index,i-index+1));
                createPartition(path,ans,i+1,s);
                path.pop_back();
            }    
        }        
    }
public:
    vector<vector<string>> partition(string s) {
        vector<string> path;
        vector<vector<string>> ans;
        createPartition(path,ans,0,s);
        return ans;
        
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.rs
fn palindrome(chars: &Vec<char>, mut start: usize, mut end: usize) -> bool {
    while start <= end && end < chars.len() {
        if chars[start] != chars[end] {
            return false;
        }
        start += 1;
        end -= 1;
    }
    return true;
}
fn create_partition(chars: &Vec<char>, ans : &mut Vec<Vec<String>>, 
    path: &mut Vec<String>, index : usize ){
        if(index>=chars.len()){
            ans.push(path.clone());
            return;
        }        
        for i in index..chars.len() {
            if palindrome(&chars,index,i){
                path.push(chars[index..=i].iter().collect());
                create_partition(chars,ans,path,i+1);
                path.pop();
            }
        }        
}
impl Solution {
    pub fn partition(s: String) -> Vec<Vec<String>> {
        let mut path : Vec<String> = Vec::new();
        let mut ans : Vec<Vec<String>> = Vec::new();
        let chars : Vec<char> = s.chars().collect();
        create_partition(&chars,&mut ans, &mut path,0);
        return ans;
    }
}